packages/sitecore-jss/src/personalization/layout-personalization-utils.ts

```diff
@@ -0,0 +1,142 @@
+import {
+  PersonalizableComponentRendering,
+  isComponentRendering,
+  isPersonalizableComponentRendering,
+  ComponentRendering,
+  PlaceholdersData,
+} from '../layout/models';
+
+export class LayoutPersonalizationUtils {
+  /**
+   * Builds the personalized fragment.
+   * @param {string} uid The unique identifier of a component.
+   * @param {Object.<string, ComponentRendering | null | undefined>} personalizedFragments The personalized fragments.
+   * @param {ComponentRendering} defaultComponent The default component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  buildPersonalizedFragment(
+    uid: string,
+    personalizedFragments: { [key: string]: ComponentRendering | null | undefined },
+    defaultComponent: ComponentRendering | null
+  ): ComponentRendering | null {
+    const personalizedFragment = personalizedFragments[uid];
+
+    if (personalizedFragment === null) {
+      return null;
+    }
+
+    if (personalizedFragment === undefined) {
```

Comment:
Sometimes it's null and sometimes it's undefined? Can you please explain how this works.

Reply:
we will add comment to the if statements
    // component is hidden so return null to render nothing
    // unable to resolve fragment for component so return default